libsolidity/lsp/LanguageServer.cpp

```
@@ -0,0 +1,354 @@
+				jsonRelated["location"] = toJson(secondaryLocation);
+				jsonDiag["relatedInformation"].append(jsonRelated);
+			}
+
+		diagnosticsBySourceUnit[*location->sourceName].append(jsonDiag);
+	}
+
+	for (string const& sourceUnitName: m_fileReader.sourceCodes() | ranges::views::keys)
+	{
+		Json::Value params;
+		params["uri"] = sourceUnitNameToClientPath(sourceUnitName);
+		params["diagnostics"] = move(diagnosticsBySourceUnit.at(sourceUnitName));
+		m_client.notify("textDocument/publishDiagnostics", move(params));
+	}
+}
+
+bool LanguageServer::run()
+{
+	while (!m_exitRequested && !m_client.closed())
+	{
+		optional<Json::Value> const jsonMessage = m_client.receive();
+		if (!jsonMessage)
+		{
+			m_client.error({}, ErrorCode::ParseError, "Error parsing JSONRPC request.");
+			continue;
+		}
+
+		try
+		{
+			string const methodName = (*jsonMessage)["method"].asString();
+			MessageID const id = (*jsonMessage)["id"];
+
+			if (auto handler = valueOrDefault(m_handlers, methodName))
+				handler(id, (*jsonMessage)["params"]);
+			else
+				m_client.error(id, ErrorCode::MethodNotFound, "Unknown method " + methodName);
+		}
+		catch (exception const& e)
+		{
+			m_client.error({}, ErrorCode::InternalError, "Unhandled exception: "s + e.what());
+		}
+	}
+	return m_shutdownRequested;
+}
+
+void LanguageServer::handleInitialize(MessageID _id, Json::Value const& _args)
+{
+	// The default of FileReader is to use `.`, but the path from where the LSP was started
+	// should not matter.
+	string rootPath("/");
+	if (Json::Value uri = _args["rootUri"])
+	{
+		rootPath = uri.asString();
+		if (!boost::starts_with(rootPath, "file://"))
+		{
+			m_client.error(_id, ErrorCode::InvalidParams, "rootUri only supports file URI scheme.");
+			return;
+		}
+		rootPath = rootPath.substr(7);
+	}
+	else if (Json::Value rootPath = _args["rootPath"])
+		rootPath = rootPath.asString();
+
+	m_fileReader.setBasePath(boost::filesystem::path(rootPath));
```

I don't think this is safe without removing all the sources from the `FileReader` first. Otherwise you might get name conflicts due to different files receiving same source unit names as other files already loaded before the change. Also an attempt to translate source unit names of those already loaded files back to paths will give you completely wrong results if you use the new base path.

It wasn't a problem so far because the CLI only uses it for initialization and never actually changes the path at runtime. It should have been non-changeable from the start and I wanted to refactor it at some point so that you can only set base path in constructor. At the very least `setBasePath()` should have an assert that the sources dict is empty.

---

Initialize should only be called once at the start.

---

Then let's add an assert that base path is empty before this line.

---

NB: If initialize is called multiple times, we should report a protocol error, not just fail an assertion.